doc/manual/source/user_guide/InSituPythonAnalysis.rst

```diff
@@ -0,0 +1,337 @@
+
+When Enzo finishes its computation in a cycle, the whole simulation pauses and starts the in situ analysis process.
+The ``CallInSitulibyt`` function in ``src/enzo/CallInSitulibyt.C`` conducts this process.
+Enzo then passes in simulation information and actual field data pointers to ``libyt``.
+This includes simulation information, like adaptive mesh grid hierarchy, parameters, field labels, etc, and actual simulation data pointers inside ``BaryonField`` array.  (Data fields are referenced, not copied, from ``libyt``'s perspective, although typically ``yt`` itself will make a copy as needed before any in-place changes occur.)
+``libyt`` will construct data structures to store simulation information and wrap these data pointers, so that they can be read and used in Python with minimum memory overhead.
+
+``libyt`` supports calling Python functions from simulation process,
+and it also supports user interface (**interactive mode** has :ref:`Interactive Python Prompt` and :ref:`Reloading Script` feature, and **jupyter kernel mode** has :ref:`Jupyter Notebook / JupyterLab UI` feature).
+We can update Python functions and probe data interactively in the UI.
+After in situ analysis is done, ``libyt`` frees resources allocated for itself, and the simulation will continue.
+
+At the end, when the simulation is shutting down, Enzo calls ``FinalizeLibytInterface`` in ``src/enzo/InitializeLibytInterface.C``, so that ``libyt`` finalizes the Python interpreters.
+
+How to Configure
+----------------
+**Some settings are hard-coded inside Enzo, we can customize it to our own needs.**
+
+General
+^^^^^^^
+
+* **What kind of test problem will it work?**
+
+  It will work with all test problems with dimensionality 1, 2, and 3 field data or with dimensionality 3 particle data.
+
+* **How to change import Python file name?**
+
+  The default Python script will be imported is ``inline.py``.
+
+  If we want to change the script name, set the file name without its extension to ``libyt_script_name`` in Enzo parameter file.
+  For example, we want to make the Python script to be ``test.py``:
+
+  ::
+
+      libyt_script_name = test
+
+
+* **How to call Python functions during simulation runtime? And what should I be aware of?**
+
+  We can call Python function using libyt API ``yt_run_Function`` and ``yt_run_FunctionArguments``. See how to use them `here <https://libyt.readthedocs.io/en/latest/libyt-api/run-python-function.html>`__.
+
+  Un-comment the code in ``src/enzo/CallInSitulibyt.C``:
+
+  ::
+
+    /* Run yt_run_Function and yt_run_FunctionArguments */
+    // if (yt_run_Function("yt_inline") != YT_SUCCESS) {
+    // 	   fprintf(stderr, "Error while running yt_run_Function and call yt_inline\n");
+    // 	   return FAIL;
+    // }
+    // if (yt_run_FunctionArguments("yt_inline_args", 1, "\'density\'") != YT_SUCCESS) {
+    //     fprintf(stderr, "Error while running yt_run_FunctionArguments and call yt_inline_args\n");
+    //     return FAIL;
+    // }
```

Comment:
This is a little inelegant; perhaps an `#IFDEF` that looks for the `libyt-yes` setting would be nicer? It's fine as is but easy to forget.

Reply:
Oh, though I see now that the `yt_run_FunctionArguments` call was more of a demo. This could be made clearer; I thought both were necessary.